### PR TITLE
Added Support for Tenor V2 Google API

### PR DESCRIPTION
**Pull request by tnoor-co**
- Tenor V1 API Support has been replaced with V2 API. You must get a new V2 API Key from Google Cloud.
- New Settings specific to Tenor have been added: Client Key, Locale & Country
- New Settings to limit infinite scrolling along with max number of search results. This will prevent API Rate Limits & misuse
- Additional Error Handling around conversion of Tenor V2 API Error Messages to readable format, catching missing API keys prior to calling API Provider, Rate Limits & Search Query Length limits
- Added variable support for `$api_provider` to include provider name in error messages.

### CODE REVIEW

**Comment by merefield** [comment]
@tnoor-co great stuff!  I've used your fork and it checks out nicely with a V1 key.

Thanks for adding the error condition, my only query would be this:

image

Is this perhaps a little 'techie', might it be better to render a more friendly error + the detail below?  A bit nit picky perhaps I'll defer to @xfalcox or team.

**Comment by tnoor-co** [comment]
Thank you @merefield ! I believe that error is from the v1 tenor endpoint, not v2 as it follows Google's standard error response. 

I had to refresh my Discourse Post (where I was testing the GIF) to clear that from cache and then clicked on GIF to view my changes.

**Comment by merefield** [comment]
@tnoor-co apologies, I was using your main branch, not your feature branch.

Having corrected that I now get a different error, but its quoting "Giphy" despite the fact I've chosen Tenor?:

image

**Comment by tnoor-co** [comment]
No worries! There is a default message for invalid API keys and it states Giphy by default. I don't think that was updated when Tenor was added to this plug in. Check this file:

- locales/en.yml

Also, the giphy_locale in settings is used for tenor as well but changing that could cause backwards compatibility.

**Comment by merefield** [comment]
thanks @tnoor-co, that might have been my fault, but would you like to improve that as part of this PR?  And in the meantime I'll grab a V2 key and test.

**Comment by tnoor-co** [comment]
Sure thing! I'll update that to a more generic error message to encompass all providers and see if I can update just the label of the locale setting.

**Comment by merefield** [comment]
Confirming that with a V2 key, the search is working nicely 🚀 thanks!

Yes please, that would be a great improvement 👍

**Comment by tnoor-co** [comment]
Made a number of changes to support both Tenor V1 and V2 endpoints along with some error handlings:

- settings: rearranged to keep giphy & tenor grouped together
- new setting: tenor api version
- new setting: v2 specific fields - country & locale
- new setting: additional error handling with two new custom errors for error_rate_limit (Giphy & Tenor) & error_search_too_long (Giphy)
- added api provider variable `$api_provider` to error messages
- added max search result limit as the API kept on getting called on scroll. This may prevent someone from hitting their rate limits. Limit is not set by default, so existing functionality won't be affected.
- added support for v2 tenor api while keeping v1 as it were

Tested with both GIPHY & Tenor V2. Tenor V1 is not tested, so I might need your help again there, @merefield :) Appreciate the insight & prompt follow-ups.

Also, I hope someone can squash all the commits into one. I may have gone overborne there.

**Comment by merefield** [comment]
Above and beyond!  thanks. 🙏 

Let's not make this any more complex though as all this needs to be maintained?

Tested on Tenor V2 & V1 for basic search only without testing rate limit and handling.

Very minor point on copy:

image

Can we lose the word 'set', as it's not really a set, it's just a single key?

I'm inclined to suggest this is good to go once that minor copy issue is resolved. @xfalcox

**Comment by tnoor-co** [comment]
I totally agree on the complexity issue.

> And perhaps the V1 code will need to be decommissioned at some point to simplify the repo and user settings?

Yes, we should definitely decommission it at some point in the near future. Updates are indeed very easy to do on Discourse, so making it visible or known would be a good course of action. Feel free to ping me down the road if we need to sunset v1 completely.

The translation change as requested is done.

**Comment by tnoor-co** [comment]
I totally agree on the complexity issue.

> And perhaps the V1 code will need to be decommissioned at some point to simplify the repo and user settings?

Yes, we should definitely decommission it at some point in the near future. Updates are very easy to do on Discourse, so making it visible or known would be a good course of action, followed by removing support for it. Feel free to ping me down the road when we need to sunset v1 completely.

The translation change as requested is done.

**Comment by tnoor-co** [comment]
@merefield @xfalcox PR is ready for review. This includes breaking changes as Tenor v1 support has been completely removed.

**Review by xfalcox — CHANGES_REQUESTED** [review]
Can you please fix this issue?

**Comment by xfalcox** [comment]
Amazing work @tnoor-co!!! Just one small lint issue before we can merge this.

**Comment by tnoor-co** [comment]
Thank you! Lint fix is in

**Review by xfalcox — APPROVED** [review]

**Comment by merefield** [comment]
Great work @tnoor-co! 🥇 I have my production key ready to go ;)

**Pull request merged**
